src/encoder/encoderopus.cpp

```diff
@@ -0,0 +1,439 @@
+        // - 4 bytes of string length
+        // - string data
+
+        // Add comment length field and data to comments "list"
+        for (int x = 0; x < 4; x++) {
+            unsigned char fieldValue = (commentBytesLength >> (x*8)) & 0xFF;
+            combinedComments.append(fieldValue);
+        }
+        combinedComments.append(commentBytes);
+
+        // Don't forget to include this comment in the overall size calculation
+        frameSize += (4 + commentBytesLength);
+        commentCount++;
+    }
+
+    // == Actual frame building ==
+    unsigned char* data = new unsigned char[frameSize]();
+    int pos = 0; // Current position
+
+    // Magic signature (8 bytes)
+    memcpy(data + pos, "OpusTags", 8);
+    pos += 8;
+
+    // Vendor string (mandatory)
+    // length field (4 bytes, little-endian) + actual string
+    // Write length field
+    for (int x = 0; x < 4; x++) {
+        data[pos++] = (vendorStringLength >> (x*8)) & 0xFF;
+    }
+    // Write string
+    memcpy(data + pos, vendorString, vendorStringLength);
+    pos += vendorStringLength;
+
+    // Number of comments (4 bytes, little-endian)
+    for (int x = 0; x < 4; x++) {
+        data[pos++] = (commentCount >> (x*8)) & 0xFF;
+    }
+
+    // Comment list (dynamic size)
+    int commentListLength = combinedComments.size();
+    memcpy(data + pos, combinedComments.constData(), commentListLength);
+    pos += commentListLength;
+
+    // Push finished tags frame to stream
+    ogg_packet packet;
+    packet.b_o_s = 0;
+    packet.e_o_s = 0;
+    packet.granulepos = 0;
+    packet.packetno = m_packetNumber++;
+    packet.packet = data;
+    packet.bytes = frameSize;
+
+    kLogger.debug() << data;
+
+    ogg_stream_packetin(&m_oggStream, &packet);
+    delete[] data;
+}
+
+void EncoderOpus::encodeBuffer(const CSAMPLE *samples, const int size) {
+    if (!m_pOpus) {
+        return;
+    }
+
+    int writeRequired = size;
+    int writeAvailable = m_pFifoBuffer->writeAvailable();
+    if (writeRequired > writeAvailable) {
+        kLogger.warning() << "FIFO buffer too small, loosing samples!"
+                          << "required:" << writeRequired
+                          << "; available: " << writeAvailable;
+    }
+
+    int writeCount = math_min(writeRequired, writeAvailable);
+    if (writeCount > 0) {
+        m_pFifoBuffer->write(samples, writeCount);
+    }
+
+    processFIFO();
+}
+
+void EncoderOpus::processFIFO() {
+    while (m_pFifoBuffer->readAvailable() >= m_readRequired) {
+        m_pFifoBuffer->read(m_pFifoChunkBuffer, m_readRequired);
+
+        int samplesPerChannel = m_readRequired / m_channels;
+        int result = opus_encode_float(m_pOpus, m_pFifoChunkBuffer, samplesPerChannel,
+                m_pOpusDataBuffer, kMaxOpusBufferSize);
+
+        if (result < 1) {
+            kLogger.warning() << "opus_encode_float failed:" << opusErrorString(result);
+            return;
+        }
+
+        ogg_packet packet;
```

Comment:
Wouldn't it be possible to reduce the number of member variable by introducing ogg_packet m_nextPacket?

Reply:
Maybe not, because the signature of ogg_stream_packetin() allows to modify the contents of the packet.